server/src/main/java/org/apache/iotdb/db/auth/AuthorizerManager.java

```diff
@@ -215,131 +245,310 @@ public void reset() throws AuthException {
     try {
       return iAuthorizer.processTakeSnapshot(snapshotDir);
     } finally {
-      snapshotLock.writeLock().unlock();
+      authReadWriteLock.writeLock().unlock();
     }
   }
 
   @Override
   public void processLoadSnapshot(File snapshotDir) throws TException, IOException {
-    snapshotLock.writeLock().lock();
+    authReadWriteLock.writeLock().lock();
     try {
       iAuthorizer.processLoadSnapshot(snapshotDir);
     } finally {
-      snapshotLock.writeLock().unlock();
+      authReadWriteLock.writeLock().unlock();
     }
   }
+
+  public TSStatus checkPath(String username, List<String> allPath, int permission)
+      throws AuthException {
+    authReadWriteLock.readLock().lock();
+    try {
+      User user = userCache.getIfPresent(username);
+      boolean status = true;
+      if (user != null) {
+        for (String path : allPath) {
+          if (!user.checkPrivilege(path, permission)) {
+            for (String roleName : user.getRoleList()) {
+              Role role = roleCache.getIfPresent(roleName);
+              // It is detected that the role of the user does not exist in the cache, indicating
+              // that the permission information of the role has changed.
+              // The user cache needs to be initialized
+              if (role == null) {
+                invalidateCache(username, "");
+                status = false;
+                break;
+              }
+              if (role.checkPrivilege(path, permission)) {
+                status = true;
+                break;
+              } else {
+                status = false;
+              }
+            }
+            // Prove that neither the user nor his role has permission
+            if (!status) {
+              break;
+            }
+          }
+        }
+      }
+      if (status) {
+        return RpcUtils.getStatus(TSStatusCode.SUCCESS_STATUS);
+      }
+      tPermissionInfoResp = ClusterAuthorizer.checkPath(username, allPath, permission);
+      if (tPermissionInfoResp.getStatus().getCode()
+          == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        userCache.get(username);
+        return tPermissionInfoResp.getStatus();
+      } else {
+        return tPermissionInfoResp.getStatus();
+      }
+    } finally {
+      authReadWriteLock.readLock().unlock();
+    }
+  }
+
+  /** Check the user */
+  public TSStatus checkUser(String username, String password) {
+    authReadWriteLock.readLock().lock();
+    try {
+      User user = userCache.getIfPresent(username);
+      if (user != null
+          && password != null
+          && AuthUtils.validatePassword(password, user.getPassword())) {
+        return RpcUtils.getStatus(TSStatusCode.SUCCESS_STATUS);
+      }
+      tPermissionInfoResp = ClusterAuthorizer.checkUser(username, password);
+      if (tPermissionInfoResp.getStatus().getCode()
+          == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        userCache.get(username);
```

Comment:
There is a concurrent issue here.

When the `userCache.get(username)` is invoked, the `tPermissionInfoResp` may be changed by another thread which is also invoking this method and doesn't hit the cache.

And from another aspect, it is better to put the `send-request` step inside the cache build method. It is a bit confusing that the `cacheUser()` reference a class variable at the beginning.

Reply:
And we also need to check the same issue for `roleCache`

Reply:
Different parameters are used for user login and check permissions. It seems that it is not possible to put the send request in the build cache

Reply:
If the send-request step is put in the cache build, the login and permission check operations use different parameters. In addition, the status information returned by the sending request seems to be unable to be obtained normally.